cloud/services/availabilitysets/availabilitysets.go

```diff
@@ -0,0 +1,140 @@
+
+// Service provides operations on azure resources
+type Service struct {
+	Scope AvailabilitySetScope
+	Client
+	resourceSKUCache *resourceskus.Cache
+}
+
+// New creates a new availability sets service.
+func New(scope AvailabilitySetScope, skuCache *resourceskus.Cache) *Service {
+	return &Service{
+		Scope:            scope,
+		Client:           NewClient(scope),
+		resourceSKUCache: skuCache,
+	}
+}
+
+// Reconcile creates or updates availability sets.
+func (s *Service) Reconcile(ctx context.Context) error {
+	ctx, span := tele.Tracer().Start(ctx, "availabilitysets.Service.Reconcile")
+	defer span.End()
+
+	if len(s.Scope.FailureDomains()) > 0 {
+		//noop if there are failure domains
+		return nil
+	}
+
+	asSku, err := s.resourceSKUCache.Get(ctx, string(compute.Aligned), resourceskus.AvailabilitySets)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get availability sets sku")
+	}
+
+	faultDomainCountStr, ok := asSku.GetCapability(resourceskus.MaximumPlatformFaultDomainCount)
+	if !ok {
+		return errors.Errorf("cannot find capability %s sku %s", resourceskus.MaximumPlatformFaultDomainCount, *asSku.Name)
+	}
+
+	faultDomainCount, err := strconv.ParseUint(faultDomainCountStr, 10, 32)
+	if err != nil {
+		return errors.Wrapf(err, "failed to determine max fault domain count")
+	}
+
+	for _, asSpec := range s.Scope.AvailabilitySetSpecs() {
+		s.Scope.V(2).Info("creating availability set", "availability set", asSpec.Name)
+
+		asParams := compute.AvailabilitySet{
+			Sku: &compute.Sku{
+				Name: pointer.StringPtr(string(compute.Aligned)),
+			},
+			AvailabilitySetProperties: &compute.AvailabilitySetProperties{
+				PlatformFaultDomainCount: pointer.Int32Ptr(int32(faultDomainCount)),
+			},
+			Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
+				ClusterName: s.Scope.ClusterName(),
+				Lifecycle:   infrav1.ResourceLifecycleOwned,
+				Name:        to.StringPtr(asSpec.Name),
+				Role:        to.StringPtr(infrav1.CommonRole),
+				Additional:  s.Scope.AdditionalTags(),
+			})),
+			Location: pointer.StringPtr(s.Scope.Location()),
+		}
+
+		_, err := s.Client.CreateOrUpdate(ctx, s.Scope.ResourceGroup(), asSpec.Name, asParams)
+		if err != nil {
+			return errors.Wrapf(err, "failed to create availability set %s", asSpec.Name)
+		}
+		s.Scope.V(2).Info("successfully created availability set", "availability set", asSpec.Name)
+	}
+
+	return nil
+}
+
+// Delete deletes availability sets.
+func (s *Service) Delete(ctx context.Context) error {
+	ctx, span := tele.Tracer().Start(ctx, "availabilitysets.Service.Delete")
+	defer span.End()
+
+	if len(s.Scope.FailureDomains()) > 0 {
+		//noop if there are failure domains
+		return nil
+	}
+
+	for _, asSpec := range s.Scope.AvailabilitySetSpecs() {
+		s.Scope.V(2).Info("deleting availability set", "availability set", asSpec.Name)
+		err := s.Client.Delete(ctx, s.Scope.ResourceGroup(), asSpec.Name)
+		if err != nil && azure.ResourceNotFound(err) {
+			// already deleted
+			return nil
```

Comment:
```suggestion
			continue
```

We should finish the loop to delete any potential other availability sets

Reply:
Ah yeah, my bad, sorry :(

Reply:
let's make sure this is covered by a unit test too

Reply:
Done.